File: c++/algorithm/TSP_완전탐색.cpp
#include <iostream>
#include <random>
#include <limits>
#include <cmath>
#include <vector>
#include <iomanip>
#define N 10
#define MAX 987654321

using namespace std;

double** initialize() {
	default_random_engine gen(43);
	normal_distribution<double> dis(N, 3*N);

	double** w = new double*[N];
	for(int i=0;i<N;i++){
		w[i] = new double[N];
	}

	for (int i=0; i<N;i++){
		w[i][i] = MAX; 
		for(int j=0; j<i; j++){
			double tmp = abs(dis(gen));
			if((int)floor(tmp) % 37 <1 )tmp =MAX;
			
			w[i][j] = tmp;
			w[j][i] = tmp;
		}
	}

	return w;
}

double shortest_path(double** w, bool* visited, vector<int>& path, double currentLength){
	if(path.size() == N )
		return currentLength + w[path[0]][path.back()];

	double ret = MAX;

	for(int i=0; i<N; i++){
		if(visited[i]) continue;

		int cur = path.back();
		path.push_back(i);
		visited[i]=true;
		ret = min(ret, shortest_path(w,visited,path, currentLength+w[cur][i]));
		visited[i]=false;
		path.pop_back();
	
	}

	return ret;
}

void print_result (double** w, double ret){
	for(int i=0; i<N; i++){
		for(int j=0; j<N; j++){
			if(w[i][j] == MAX) {
				cout << setw(5) << "-";
				cout << " ";
				continue;
			}
			cout <<fixed<<setw(5) << setprecision(2)<< w[i][j];
			cout << " ";
		}
		cout << endl;
	}

	cout << "result :  "<< fixed<< setprecision(2) << ret << endl;
}

int main(void){

	vector<int> path;
	double** w = initialize();
	bool visited[N];
	fill(visited,visited+N,false);
	path.push_back(0);
	double ret = shortest_path(w, visited, path, 0.);

	print_result(w,path,ret);


	return 0;
}

File: javascript/algorithm/TSP_완전탐색.mjs
const TEST_CASES = 10
const N = 10
const MAX = 987654321
const MAX_WEIGHT = 100


function initialize(){
	const ret = new Array(N).fill(0)
				.map(v => new Array(N).fill(MAX))

	for (let i=0; i<N; i++) {
		for(let j= i+1; j<N; j++){
			const tmp = Math.floor(Math.random() * MAX_WEIGHT) + 1
			if (tmp % 13 == 0 ) continue
			ret[i][j] = tmp
			ret[j][i] = tmp
		}
	}

	return ret 
}

function calculate(w, path) {
	let ret = 0
	let start, end

	for (let i =1; i < N; i++) {
		start = path[i-1]
		end = path[i]
		ret += w[start][end]
	}

	start = path[0]
	end = path[N-1]

	return w[end][start] == MAX ? MAX : ret + w[end][start]
}

function search(w, visited, path) {
	if (path.length == N) {
		return calculate(w, path)
	}

	let ret = Number.MAX_VALUE
	const cur = path[path.length-1]

	for (let i=0; i<N; i++) {
		if(visited[i] || w[cur][i] == MAX) continue

		visited[i] = true
		path.push(i)
		ret = Math.min(search(w,visited,path), ret)
		path.pop()
		visited[i] = false
	}

	return ret
}

function solve(w) {
	const visited = new Array(N).fill(false)
	visited[0] = true
	return search(w, visited, [0])
}


for (let it=0; it< TEST_CASES; it++) {
	const w = initialize()
	console.log(solve(w))
}

